src/components/FloatingToolbar.tsx

```diff
@@ -1,170 +1,113 @@
-    const range = getMarkRange(selection.$from, state.schema.marks.link);
-
-    let items = [];
-    if (isTableSelection) {
-      items = getTableMenuItems();
-    } else if (colIndex !== undefined) {
-      items = getTableColMenuItems(state, colIndex);
-    } else if (rowIndex !== undefined) {
-      items = getTableRowMenuItems(state, rowIndex);
-    } else {
-      items = getFormattingMenuItems(state);
+    // based on the start and end of the selection calculate the position at
+    // the center top
+    const startPos = view.coordsAtPos(selection.$from.pos);
+    const endPos = view.coordsAtPos(selection.$to.pos);
+
+    // tables are an oddity, and need their own logic
+    const isColSelection =
+      selection.isColSelection && selection.isColSelection();
+    const isRowSelection =
+      selection.isRowSelection && selection.isRowSelection();
+
+    if (isRowSelection) {
+      endPos.left = startPos.left + 12;
+    } else if (isColSelection) {
+      const { node: element } = view.domAtPos(selection.$from.pos);
+      const { width } = element.getBoundingClientRect();
+      endPos.left = startPos.left + width;
     }
 
-    if (!items.length) {
-      return null;
-    }
+    const halfSelection = Math.abs(endPos.left - startPos.left) / 2;
+    const centerOfSelection = startPos.left + halfSelection;
+
+    // position the menu so that it is centered over the selection except in
+    // the cases where it would extend off the edge of the screen. In these
+    // instances leave a margin
+    const { offsetWidth, offsetHeight } = this.menuRef.current;
+    const margin = 12;
+    const left = Math.min(
+      window.innerWidth - offsetWidth - margin,
+      Math.max(margin, centerOfSelection - offsetWidth / 2)
+    );
+    const top = Math.min(
+      window.innerHeight - offsetHeight - margin,
+      Math.max(margin, startPos.top - offsetHeight)
+    );
+
+    // if the menu has been offset to not extend offscreen then we should adjust
+    // the position of the triangle underneath to correctly point to the center
+    // of the selection still
+    const offset = left - (centerOfSelection - offsetWidth / 2);
+
+    console.log({ left, centerOfSelection, offsetWidth, offset });
+
+    return {
+      left: left + window.scrollX,
+      top: top + window.scrollY,
+      offset,
+    };
+  }
+
+  render() {
+    const { children, active } = this.props;
 
     return (
       <Portal>
         <Wrapper
-          active={isActive}
-          ref={menuRef}
+          active={active}
+          ref={this.menuRef}
           top={this.state.top}
           left={this.state.left}
           offset={this.state.offset}
         >
-          {link && range ? (
-            <LinkEditor
-              mark={range.mark}
-              from={range.from}
-              to={range.to}
-              {...this.props}
-            />
-          ) : (
-            <Menu items={items} {...this.props} />
-          )}
+          {children}
```

Comment:
- somehow this file confuses the heck out of git
- I removed the business logic and left `FloatingToolbar` as the display wrapper